pkg/gc/config.go

```diff
@@ -81,41 +85,79 @@ func NewConfigFromConfigMapFunc(ctx context.Context) func(configMap *corev1.Conf
-			cm.AsDuration("retain-since-last-active-time", &c.RetainSinceLastActiveTime),
+			// v2 settings
+			cm.AsString("retain-since-create-time", &retainCreate),
+			cm.AsString("retain-since-last-active-time", &retainActive),
 			cm.AsInt64("min-stale-revisions", &c.MinStaleRevisions),
-			cm.AsInt64("max-stale-revisions", &c.MaxStaleRevisions),
+			cm.AsString("max-non-active-revisions", &max),
 		); err != nil {
 			return nil, fmt.Errorf("failed to parse data: %w", err)
 		}
 
-		if c.MaxStaleRevisions >= 0 && c.MinStaleRevisions > c.MaxStaleRevisions {
-			return nil, fmt.Errorf(
-				"min-stale-revisions(%d) must be <= max-stale-revisions(%d)",
-				c.MinStaleRevisions, c.MaxStaleRevisions)
-		}
-		if c.MinStaleRevisions < 0 {
-			return nil, fmt.Errorf("min-stale-revisions must be non-negative, was: %d", c.MinStaleRevisions)
-		}
-		if c.MaxStaleRevisions < -1 {
-			return nil, fmt.Errorf("max-stale-revisions must be >= -1, was: %d", c.MaxStaleRevisions)
-		}
-
 		if c.StaleRevisionMinimumGenerations < 0 {
-			return nil, fmt.Errorf("stale-revision-minimum-generations must be non-negative, was: %d", c.StaleRevisionMinimumGenerations)
+			return nil, fmt.Errorf(
+				"stale-revision-minimum-generations must be non-negative, was: %d",
+				c.StaleRevisionMinimumGenerations)
 		}
-
 		if c.StaleRevisionTimeout-c.StaleRevisionLastpinnedDebounce < minRevisionTimeout {
 			logger.Warnf("Got revision timeout of %v, minimum supported value is %v",
 				c.StaleRevisionTimeout, minRevisionTimeout+c.StaleRevisionLastpinnedDebounce)
 			c.StaleRevisionTimeout = minRevisionTimeout + c.StaleRevisionLastpinnedDebounce
 		}
+
+		// validate V2 settings
+		if err := parseDisabledOrDuration(retainCreate, &c.RetainSinceCreateTime); err != nil {
+			return nil, fmt.Errorf("failed to parse min-stale-revisions: %w", err)
+		}
+		if err := parseDisabledOrDuration(retainActive, &c.RetainSinceLastActiveTime); err != nil {
+			return nil, fmt.Errorf("failed to parse retain-since-last-active-time: %w", err)
+		}
+		if err := parseDisabledOrInt64(max, &c.MaxNonActiveRevisions); err != nil {
+			return nil, fmt.Errorf("failed to parse max-stale-revisions: %w", err)
+		}
+		if c.MinStaleRevisions < 0 {
+			return nil, fmt.Errorf("min-stale-revisions must be non-negative, was: %d", c.MinStaleRevisions)
+		}
+		if c.MaxNonActiveRevisions >= 0 && c.MinStaleRevisions > c.MaxNonActiveRevisions {
+			return nil, fmt.Errorf("min-stale-revisions(%d) must be <= max-stale-revisions(%d)", c.MinStaleRevisions, c.MaxNonActiveRevisions)
+		}
 		return c, nil
 	}
 }
+
+func parseDisabledOrInt64(val string, toSet *int64) error {
+	if val == "" {
+		// keep default value
+	} else if strings.EqualFold(val, disabled) {
+		*toSet = Disabled
+	} else if parsed, err := strconv.ParseInt(val, 10, 64); err != nil {
+		return err
+	} else if parsed < 0 {
+		return fmt.Errorf("must non-negative or %q, was: %d", disabled, parsed)
+	} else {
+		*toSet = parsed
+	}
+	return nil
+}
+
```

Comment:
```suggestion
switch {
case val == "":
case strings.EqualFold(val, disabled):
 *toSet = Disabled
default:
 parsed, err := strconv.ParseUint(val, 10, 64)
 if err != nil {
		return err
 }
 *toSet = parsed
}
return nil
}

```
Something like that I think reads better.
Same below.